File: web/src/services/apiService.js
const BASE_URL = "http://localhost:8080/api";

// helper function completely hidden from the UI components
function getHeaders() {
  const token = localStorage.getItem("trackademia_token");
  return {
    "Content-Type": "application/json",
    ...(token ? { "Authorization": `Bearer ${token}` } : {})
  };
}

// the facade
export const apiService = {
  
  login: async (email, password) => {
    const response = await fetch(`${BASE_URL}/auth/login`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email, password })
    });
    if (!response.ok) throw new Error("Login failed");
    return response.json();
  },

  getTasks: async () => {
    const response = await fetch(`${BASE_URL}/tasks`, { headers: getHeaders() });
    if (!response.ok) throw new Error("Failed to fetch tasks");
    return response.json();
  },

  createTask: async (taskData) => {
    const response = await fetch(`${BASE_URL}/tasks`, {
      method: "POST",
      headers: getHeaders(),
      body: JSON.stringify(taskData)
    });
    if (!response.ok) throw new Error("Failed to create task");
    return response.json();
  },

  completeTask: async (taskId) => {
    const response = await fetch(`${BASE_URL}/tasks/${taskId}/complete`, {
      method: "PUT",
      headers: getHeaders()
    });
    if (!response.ok) throw new Error("Failed to update task");
    return response.json();
  }
};

File: mobile/app/src/main/java/com/example/trackademia/api/ApiService.kt
package com.example.trackademia.api

import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path

interface ApiService {
    @POST("auth/login")
    suspend fun login(@Body request: LoginRequest): Response<AuthResponse>

    @POST("auth/register")
    suspend fun register(@Body request: RegisterRequest): Response<AuthResponse>

    @GET("tasks")
    suspend fun getTasks(): Response<List<Task>>

    @POST("tasks")
    suspend fun createTask(@Body task: TaskRequest): Response<Task>

    @PUT("tasks/{taskId}")
    suspend fun updateTask(@Path("taskId") taskId: Int, @Body task: TaskRequest): Response<Task>

    @PUT("tasks/{taskId}/complete")
    suspend fun completeTask(@Path("taskId") taskId: Int): Response<Task>

    @GET("activities")
    suspend fun getActivities(): Response<List<ActivityLog>>
}
